include/swift/SIL/match.h

```diff
@@ -0,0 +1,200 @@
+//===--- Match.h - Range based matching function -------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the source for a range based matching function. The function takes a begin and end iterator
+/// which it will use to try to match each element agains a corresponding parameter.
+///
+//===----------------------------------------------------------------------===//
+
+
+#ifndef SWIFT_SIL_MATCH_H
+#define SWIFT_SIL_MATCH_H
+
+#include "swift/SIL/SILInstruction.h"
+#include "llvm/ADT/Optional.h"
+
+namespace swift {
+
+template <template <int...> class _Tp, class>
+struct instance_of : std::false_type {};
+
+template <template <int...> class _Tp, int... _Args>
+struct instance_of<_Tp, _Tp<_Args...>> : std::true_type {};
+
+/// InfoBlock is a struct that is used to give info about the element you want to match.
+/// \p T is the type that you want to match. The match function will use
+/// \code dyn_cast \endcode to try to cast the current element to this type. \p R is
+/// a boolean which tells the matching function weather you want the element that you have just
+/// matched back. If this is true, that element will be added to the returned tuple.
+/// \p L is an optional lambda type. If the dynamic cast succeeds the lambda will be passed the
+/// element. If the lambda returns true, the matching succeeds otherwise the matching fails.
+template <class T, bool R, class L = char>
+struct InfoBlock {
+  typedef L pred_t;
+  typedef T type;
+  static constexpr inline bool value = R;
+
+  L lambda;
+
+  InfoBlock(L lambda = 0) : lambda(lambda) {}
+};
+
+/// make_info_block is a helper function for creating an InfoBlock.
+/// make_info_block should be used when you want to supply an optional
+/// lambda to the InfoBlock. \p lambda is the lambda, all other
+/// arguments are the same as InfoBlock.
+template <class T, bool R, class L = char>
+static InfoBlock<T, R, L> make_info_block(L lambda = 0) {
+  return InfoBlock<T, R, L>(lambda);
+}
+
+struct Dummy {
+  static bool classof(const ValueBase *) { return true; };
+  static bool classof(const SILNode *) { return true; }
+};
+
+/// A range can be used to ignore a certain number of elements. The range will
+/// ignore at least \p Begin elements and will ignore up to \p End elements.
+template <int Begin = 0, int End = std::numeric_limits<int>::max()>
+struct Range {
+  static constexpr inline bool value = false;
```

Comment:
`inline` member variables for LinkonceODR is a C++17 thing.  We're stuck on 14, unfortunately.

Reply:
Are we allowed to use C++17 extensions? I don't see any reason why not. Swift always compiles with clang, right?

Reply:
If we're able to express those extensions (usually newfangled stdlib utilities) in C++14, we will (see llvm ADT).  But any language constructs need to backwards-deploy.

Reply:
What's the minimum clang version Swift supports? Could you point me towards the build requirements file? Clang has supported constexpr if statements (for example) as a C++14 extension since version 3.9.

Reply:
It's a function of downstream dependencies as well.  [See this thread](https://forums.swift.org/t/switching-swift-to-c-14/7251/7) for when we switched from 11 to 14.

Reply:
Thanks for the link. It seems Swift requires at minimum Clang 3.6 which means these extensions won't work. Maybe I'll start another thread trying to update this so we can get some C++17 extensions without upgrading to C++17 all together.